server/src/services/check.service.ts

```diff
@@ -0,0 +1,95 @@
+import { getRepository } from 'typeorm';
+import { CourseTask, Student, Task, TaskSolutionResult, User } from '../models';
+
+const LOW_ERROR_RATE = 0.9;
+const HIGH_ERROR_RATE = 1.1;
+const TASK_REVIEW_COUNT = 4;
+const MIN_LENGTH_MESSAGE = 70;
+
+/* Get checkers who passed max score for everyone and maybe didn't review task*/
+export async function getCheckersWithMaxScore(taskId: number) {
+  const data = await getRepository(TaskSolutionResult)
+    .createQueryBuilder('ts')
+    .select('t.name', 'taskName')
+    .addSelect('"checkerUser"."githubId"', 'checkerGithubId')
+    .addSelect('"studentUser"."githubId"', 'studentGithubId')
+    .addSelect('"studentAvg"."avg"', 'studentAvgScore')
+    .addSelect('ts.score', 'checkerScore')
+    .innerJoin(
+      qb =>
+        qb
+          .subQuery()
+          .select('tsr."studentId"')
+          .addSelect('AVG(tsr.score)', 'avg')
+          .from(TaskSolutionResult, 'tsr')
+          .groupBy('tsr."studentId"'),
+      'studentAvg',
+      'ts."studentId" = "studentAvg"."studentId"',
+    )
+    .innerJoin(CourseTask, 'ct', 'ts."courseTaskId" = ct.id')
+    .innerJoin(Task, 't', 'ct."taskId" = t.id')
+    .innerJoin(Student, 'checker', 'ts."checkerId" = checker.id ')
+    .innerJoin(User, 'checkerUser', 'checker."userId" = "checkerUser".id')
+    .innerJoin(Student, 'student', 'ts."studentId" = student.id ')
+    .innerJoin(User, 'studentUser', 'student."userId" = "studentUser".id')
+    .where(qb => {
+      const subQuery = qb
+        .subQuery()
+        .select('checkers."checkerId"')
+        .from(qb => {
+          return qb
+            .select('ts."checkerId"')
+            .addSelect('COUNT(score)', 'count')
+            .addSelect('AVG(score)', 'avg')
+            .from(TaskSolutionResult, 'ts')
+            .groupBy('ts."checkerId"')
+            .addGroupBy('ts."score"')
+            .having(`COUNT(score) = ${TASK_REVIEW_COUNT}`);
```

Comment:
Never do like that. It can leed to SQL injections in the future.
```
.having('COUNT(score) = :count', { count: TASK_REVIEW_COUNT }); 
```
but check the syntax it's probably just
```
.having('COUNT(score) = $1', [TASK_REVIEW_COUNT]); 
```
I'm not sure...

Reply:
I don't know why, but in the first time with parametrs object i had an error(
Know i didn't have error
Fixed